flink-autoscaler/src/main/java/org/apache/flink/autoscaler/JobVertexScaler.java

```diff
@@ -389,15 +407,61 @@ protected static int scale(
             return newParallelism;
         }
 
-        // When the shuffle type of vertex inputs contains keyBy or vertex is a source, we try to
-        // adjust the parallelism such that it divides the maxParallelism without a remainder
-        // => data is evenly spread across subtasks
-        for (int p = newParallelism; p <= maxParallelism / 2 && p <= upperBound; p++) {
-            if (maxParallelism % p == 0) {
+        int numKeyGroupsOrPartitions = maxParallelism;
+        int upperBoundForAlignment;
+        if (numPartitions <= 0) {
+            upperBoundForAlignment = Math.min(maxParallelism / 2, upperBound);
+        } else {
+            upperBoundForAlignment = Math.min(numPartitions, upperBound);
```

Comment:
Should the source vertex respect the `maxParallelism`?

If maxParallelism is 100, source has 1000 partition, and `upperBound` is 200. The new parallelism may be greater than 100, right? 

IIUC, the flink job won't run when `parallelism > maxParallelism` even if without key group.

Reply:
upperBound is already the smaller value obtained by comparing parallelismUpperLimit and maxParallelism.